linux/mcompile/keymap/mc_import_rules.cpp

```diff
@@ -0,0 +1,651 @@
+        if((kkp->ShiftFlags & (K_CTRLFLAG|K_ALTFLAG|LCTRLFLAG|LALTFLAG|RCTRLFLAG|RALTFLAG)) != 0) {
+          gp2->cxKeyArray++;
+          LPKMX_KEY kkp2 = new KMX_KEY[gp2->cxKeyArray];
+          memcpy(kkp2, gp2->dpKeyArray, sizeof(KMX_KEY)*(gp2->cxKeyArray-1));
+          gp2->dpKeyArray = kkp2;
+          kkp2 = &kkp2[gp2->cxKeyArray-1];
+          kkp2->dpContext = new KMX_WCHAR; 
+          *kkp2->dpContext = 0;
+          kkp2->Key = kkp->Key;
+          kkp2->ShiftFlags = kkp->ShiftFlags;
+          kkp2->Line = 0;
+          KMX_WCHAR *p = kkp2->dpOutput = new KMX_WCHAR[4];
+          *p++ = UC_SENTINEL;
+          *p++ = CODE_USE;
+          *p++ = (KMX_WCHAR)(kp->cxGroupArray);
+          *p = 0;
+        }
+      }
+    }
+  }
+
+  // If we have deadkeys, then add a new group to translate the deadkeys per the deadkey tables
+  // We only do this if not in deadkey conversion mode
+  //
+
+  if (alDead.size() > 0 && !bDeadkeyConversion) {   // I4552
+    kp->cxGroupArray++;
+
+    KMX_WCHAR *p = gp->dpMatch = new KMX_WCHAR[4];
+    *p++ = UC_SENTINEL;
+    *p++ = CODE_USE;
+    *p++ = (KMX_WCHAR) kp->cxGroupArray;
+    *p = 0;
+
+    gp++;
+
+    gp->fUsingKeys = FALSE;
+    gp->dpMatch = NULL;
+    gp->dpName = NULL;
+    gp->dpNoMatch = NULL;
+    gp->cxKeyArray = alDead.size();
+    LPKMX_KEY kkp = gp->dpKeyArray = new KMX_KEY[alDead.size()];
+
+    LPKMX_STORE sp = new KMX_STORE[kp->cxStoreArray + alDead.size() * 2];
+    memcpy(sp, kp->dpStoreArray, sizeof(KMX_STORE) * kp->cxStoreArray);
+
+    kp->dpStoreArray = sp;
+
+    sp = &sp[kp->cxStoreArray];
+    int nStoreBase = kp->cxStoreArray;
+    kp->cxStoreArray += alDead.size() * 2;
+
+    for(UINT i = 0; i < alDead.size(); i++) {
+      DeadKey *dk = alDead[i];
+
+      sp->dpName = NULL;
+      sp->dwSystemID = 0;
+      sp->dpString = new KMX_WCHAR[dk->KMX_Count() + 1];
+      for(int j = 0; j < dk->KMX_Count(); j++)
+        sp->dpString[j] = dk->KMX_GetBaseCharacter(j);
+      sp->dpString[dk->KMX_Count()] = 0;
+      sp++;
+
+      sp->dpName = NULL;
+      sp->dwSystemID = 0;
+      sp->dpString = new KMX_WCHAR[dk->KMX_Count() + 1];
+      for(int j = 0; j < dk->KMX_Count(); j++)
+        sp->dpString[j] = dk->KMX_GetCombinedCharacter(j);
+      sp->dpString[dk->KMX_Count()] = 0;
+      sp++;
+
+      kkp->Line = 0;
+      kkp->ShiftFlags = 0;
+      kkp->Key = 0;
+      KMX_WCHAR *p = kkp->dpContext = new KMX_WCHAR[8];
+      *p++ = UC_SENTINEL;
+      *p++ = CODE_DEADKEY;
+      *p++ = KMX_DeadKeyMap(dk->KMX_DeadCharacter(), &alDead, nDeadkey, FDeadkeys);   // I4353
+      // *p++ = nDeadkey+i;
+      *p++ = UC_SENTINEL;
+      *p++ = CODE_ANY;
+      *p++ = nStoreBase + i*2 + 1;
+      *p = 0;
+
+      p = kkp->dpOutput = new KMX_WCHAR[5];
+      *p++ = UC_SENTINEL;
+      *p++ = CODE_INDEX;
+      *p++ = nStoreBase + i*2 + 2;
+      *p++ = 2;
+      *p = 0;
+      kkp++;
+    }
+  }
+
+return true;
+}
+
+const int CODE__SIZE[] = {
```

Comment:
Is this used anywhere?

Reply:
Might be better to move it next to the implementation of `KMX_incxstr` in `mc_kmxfile.cpp` then. That would also allow to remove the declaration from `kmx_file.h`.

Reply:
OK-changed- see PR #11816;
also #included kmx_file.h from common instead of having a copy in mcompile